File: 4clojure/077-anagrams.hs
#!/usr/bin/env runghc

import Data.List (sort)
import Data.Map.Strict (insertWith, elems, empty)

addWord memo word = insertWith (++) (sort word) [word] memo
anagrams words = elems $ foldl addWord empty words

main = print $ anagrams ["veer", "lake", "item", "kale", "mite", "ever"]

File: 4clojure/077-anagrams.js
#!/usr/bin/env node

const groupAnagram = (memo, word) => {
  const key = word.split('').sort().join('')
  memo[key] ? memo[key].add(word) : memo[key] = new Set([word])
  return memo
}

const values = (obj) =>
  Object.keys(obj).map(key => obj[key])


const anagrams = (words) =>
  values(words.reduce(groupAnagram, {}))


console.log(anagrams(['veer', 'lake', 'item', 'kale', 'mite', 'ever']))
